### ISSUE TITLE
💡 RFC: Finalize component script scope/behavior

### ISSUE THREAD

**Issue by FredKSchott** [snowpackjs/astro#619]
## Motivation

Up to now, the component frontmatter script has been defined at the component level. This means that all logic runs on every render. To use React as an analogy, it is as if the component script lives inside of the component render function.

This has caused trouble when some expensive logic lives at the component level. For example, data fetching inside of an Astro component would happen multiple times if a component lived on the page in multiple places.

Svelte solves this problem by having a special, second script type: `<script context="module">`. We would like to keep our scripting to a single block, since having multiple `---` blocks would break our current understanding of how frontmatter looks/feels.

## Proposed Solution

_The goal of this RFC is to finalize/solidify this behavior, and give clear documentation for how to do certain things that are currently hard or not supported today._

___NOTE!!!__ A second design is being explored in the comments. Both designs are being weighed equally._

### 1. The frontmatter script is always instance-level. 
- This is a commitment to current behavior.
- Generally speaking, the frontmatter script will run for every component render

### 2b. Exported things are hoisted out of the component scope.
- The svelte compiler that we based our compiler on already does this sort of hoisting today. 
- You will get an error if something exported is impure (for example, if an exported function references any props or other variables outside of itself).
- Referencing imports inside of exported functions is supported, since imports are also hoisted out.
- TBD: can an exported function reference an exported variable? This would be technically impure, but also technically supported due to how hoisting would work.

### 3. To share state across components, use a store (you can also export a store)
- This would give the ability to support some shared state across components, if they all reference the same store.
- This relies on a TBD store API.

### 4. Perf considerations
components are currently async, and you can do a top-level async await. By committing to an instance-level component script, we must also commit to solving the potential perf issues that come with this.
- Automatic warnings if a component takes more than a few ms to render AND is rendered multiple times in a build
- respect cache headers with `fetch()` to automatically cache API responses
- `import {defineOnce} from 'astro/util'` to support memoization of values

### 5. What If none of the above supports your use-case?
- You can still always define your JS code outside of the component. For instance, a `foo.module.js` sibling

## Alternative Designs
- Vue-style component instances: https://github.com/snowpackjs/astro/issues/619#issuecomment-876044245

**Comment by matthewp** [comment]
I think __2a__ and __2b__ are pretty problematic. We can document that you must not rely on scope variables when defining exports but people are going to do so anyways and not read the docs. It just goes against expectations to say that a block of JavaScript doesn't obey the normal rules of JavaScript closures.

That's why I feel that *some* explicit solution is required here. I think it's perfectly fine for us to invent new syntax extension on top of the frontmatter `---` syntax. We're *already* extending frontmatter syntax since JavaScript is not recognized by any frontmatter parsers (at least any I'm aware of).

**Comment by FredKSchott** [comment]
> You will get an error if something exported is impure (for example, if an exported function references any props or other variables outside of itself).

Does this line make you feel differently? The idea is that you would get a very clear error message if you did this, with a link to the docs. Our compiler would basically guarantee this point for you.

I disagree that this is breaking any user expectations on how JavaScript works. We're not saying anything more than "exported functions must be pure". If we allowed impure exported functions then I agree, this would be a bit of a mess. But because this proposal is just to disallow impure exports entirely, we aren't changing anything about how you would expect your JS script to run. To the user, this would just be a restriction enforced by the compiler.

**Comment by eyelidlessness** [comment]
> Up to now, the component frontmatter script has been defined at the component level. This means that all logic runs on every render. To use React as an analogy, it is as if the component script lives inside of the component render function.
> …
> - This is a commitment to current behavior.
> - Generally speaking, the frontmatter script will run for every component render

This is *quite* surprising, and suggests that the frontmatter block is semantically analogous to a script tag rather than an ESM module. But that violates the expectation one would have when importing an Astro component from a real ESM module.

I believe this is why Svelte and Vue SFCs use explicit script tags. Or at least it makes this sort of behavior significantly less surprising. Of course JSX frameworks don’t need to address this at all because modules always follow module semantics.

> You will get an error if something exported is impure (for example, if an exported function references any props or other variables outside of itself).
> …
> TBD: can an exported function reference an exported variable? This would be technically impure, but also technically supported due to how hoisting would work.

Neither of these cases are necessarily impure. Referencing a constant primitive or calling a non-exported pure function is still guaranteed to be pure. Referencing a non-primitive is only impure if something mutates it.

> Referencing imports inside of exported functions is supported, since imports are also hoisted out.

I’m unclear on what “hoisted” means here, but I’m assuming that means it’s treated as a normal ESM import, which is basically memoized. If that’s right, I’m concerned that this is a back door to making exports impure. Sure, the import itself should only occur once, but it can in turn export impure functions.

Regarding 2a/2b generally, I think the underlying motivation may be better served by memoization rather than rules about purity. Determining with confidence whether an arbitrary function is actually pure in JS is likely impossible, and easy to get wrong.

> This relies on a TBD store API.

Is there a compelling reason to reinvent this particular wheel? People are likely already using a state management solution they prefer.

**Comment by kolodziejczak-sz** [comment]
Hey guys I think it's worth to take a look how elder.js and eleventy solves this problem:

elder js https://elderguide.com/tech/elderjs/#6-the-data-hook-is-executed
eleventy https://www.11ty.dev/docs/data-template-dir/

TLDR:
Elder.js uses the hooks to inject the data, the `data` object is accesible across all components
Eleventy API allows to create a dedicate files for fetching the data or you can use the `addGlobalData` fn

**Comment by FredKSchott** [comment]
> This is quite surprising, and suggests that the frontmatter block is semantically analogous to a script tag rather than an ESM module. But that violates the expectation one would have when importing an Astro component from a real ESM module.

We actually introduced the frontmatter script for exactly this reason: separating the component "script" from the "template". If we move this to a `<script>` tag, we now have the idea of our HTML template defining behavior + markup, instead of just markup. I agree though that it's not as cut-and-dry as "oh, this is the actual JS script".

If we went that route and said that the component frontmatter script is actually at the file level, then we'd still need some way to set props & behavior at the component level. I wonder what that could look like? Maybe something more like Vue 3's setup script? 
- https://learnvue.co/2021/05/explaining-the-new-script-setup-type-in-vue-3-major-takeaways-from-the-rfc/ 
- https://github.com/vuejs/rfcs/blob/13695db1c3bc44cb929f457cfd622526d1f4fbed/active-rfcs/0040-script-setup.md

Or maybe just Vue's less experimental composition API, where a component script can `export default` some composition API that defines actual component instance behavior? I do think that the `script setup` API is too advanced for most users (with `ref`, `computed`, etc) so this could be a good, well-proven path for us to take.

- https://v3.vuejs.org/api/composition-api.html#setup

**Comment by FredKSchott** [comment]
> Hey guys I think it's worth to take a look how elder.js and eleventy solves this problem:
> 
> elder js https://elderguide.com/tech/elderjs/#6-the-data-hook-is-executed
> eleventy https://www.11ty.dev/docs/data-template-dir/

Curious how we could map this into Astro! I think this would mean that components can't do their own data loading, and instead you should do all data loading in some central place? Maybe only `pages/` support top-level await/fetch and then `components/` are much, much simpler and UI-only?

**Comment by FredKSchott** [comment]
> Or maybe just Vue's less experimental composition API, where a component script can export default some composition API that defines actual component instance behavior? I do think that the script setup API is too advanced for most users (with ref, computed, etc) so this could be a good, well-proven path for us to take.

Okay, I wanted to explore this a bit: What if we went the opposite direction as OP, and considered the component frontmatter script a file-level scope. We would then need to move any component-level logic into a default export. Something like this:

```jsx
---
// This is the file-level scope. Do whatever you want here! Import, export, etc etc!
export function foo() { /* ... */ }
const posts = await fetch('...');

// You should still define your props using TS:
export interface Props {
  title: string;
  message?: string;
}

// This is the component-level behavior
export default {
  // You still need a place to define default props:
  props({props}) {
    return {...props, message: "Default message"};
  },
  // Define your component-level data here:
  async data({props}) {
    return {
      componentScopedPosts: await fetch('...'),
      someComputedThing: data.posts.length,
    }
  },
}
---
<div>
  <h1>Prop: {$props.title} and also {$props.message}</h1>
  <h2>File-scope variable: {posts.length}</h1>
  <h3>Component-scope data: {$data.someComputedThing} and also {$data.componentScopedPosts.length}</h1>
</div>
```

This is a bit more verbose and a bit more boilerplate as you move into more complex use-cases. However, it's much more explicit about where logic lives and when different logic executes, which is nice. Another nice thing is that you only need to opt into this complexity as you need it. So static Astro components would still be very simple, and you only need to reach for this API when you need props, or data loading.

#### What about pages?

I think that pages would never need this API: since they get no props and are only ever rendered once. A few points though:
- Things like `Astro.request` would probably need to become magic variables like `$request` that are accessible in the template and component context (`export default {}`) only. Similar to how we turn props into `$props` in the template example above.
- The collections API could take advantage of this, instead of being a magic `export` function that gets hoisted, it is a function defined in a page's component context (`export default {}`).

@eyelidlessness @matthewp @snowpackjs/contributors-core thoughts?

**Comment by kolodziejczak-sz** [comment]
> 
> 
> > Hey guys I think it's worth to take a look how elder.js and eleventy solves this problem:
> > elder js https://elderguide.com/tech/elderjs/#6-the-data-hook-is-executed
> > eleventy https://www.11ty.dev/docs/data-template-dir/
> 
> Curious how we could map this into Astro! I think this would mean that components can't do their own data loading, and instead you should do all data loading in some central place? Maybe only `pages/` support top-level await/fetch and then `components/` are much, much simpler and UI-only?

Yup, the routes are meant to be created in a simple manner where the data-provider and view are separated.
Elder.js has a similar API to the createCollection fn but the function itself is meant to be stored in a dedicated file called route.js and is available only to pages (and hooks), not the components.

That being said I think it's a nice idea to have some constants params that are being injected to every page or component. Of course, we would have to provide a way to define the values for them. Currently the createCollection function does that with params.collection.
This might be done by hooks or another tailored function, like you already referenced the setup fn from vue3.

But, there's also another case where it would be nice to reduce the network requests: having at least two routes that use the same fetched data. In elder.js you can use the bootstrap hook to inject the fetched data to the global scope and reuse it.

I guess my answer lead to conclusion that having the space to provide globals before any page execution seems very attractive for developer, and I would love to see that kind of option in astro.js.

**Comment by matthewp** [comment]
> Does this line make you feel differently? The idea is that you would get a very clear error message if you did this, with a link to the docs. Our compiler would basically guarantee this point for you.

Not really, this isn't so much about purity as it is about using the scope. You wouldn't be able to do this:

```js
const SOME_VALUE_THAT_DOENST_CHANGE = 22;

export function calc() {
  return SOME_VALUE_THAT_DOENST_CHANGE;
}
```

Because the exported function is moved to a different scope. This defies my expectation when looking at JavaScript. I suppose a smart enough compiler could moved used variables into the right scope. That seems a little difficult to implement and I'm not sure there aren't cases where it hoists things we wouldn't want it to (but I could be wrong).

**Comment by FredKSchott** [comment]
Okay, I played around a bit with the Vue-like API, and here's the best that I could come up with. You can run this yourself in the `portfolio` example of this branch/commit (looks a little different there, but functionally the same): https://github.com/snowpackjs/astro/commit/59728dd64a4586d729a31d5b0daae2e4bd532c0d

```jsx
---
export interface Props {
    title: string;
    description?: string[];
}

// If you need more than what the props give you, add a `data()` fn
// use this for: 1. default prop, 2. Astro.fetchContent calls, 3. a full await-fetch call, 4. etc.
export function data(props: Props) {
    return {description: props.description || 'This is the default site descrition.'};
}
---
<meta charset="UTF-8">
<title>{$props.title}</title>
<meta name="description" content={$data.description}>
```

#### What I like about this:
- Scoped very simply to just "prepare component-specific data". Everything else is now file-level.
- This pattern would set us up well to add something very similar to `getStaticProps()`, etc. for dynamic pages.
- No compiler magic. This lets us remove all (literally 100%) of our current script AST modifications.

I had originally tried `export default { ... }` and some idea of `defaultProps` but then realize that this really doesn't work with TypeScript. For ex, with `defaultProps` a value like `description` could be always defined inside the template, but allowed to be optional in the exported `Props` interface. It is impossible to represent both with a single `Props` interface.

#### What I dislike about this:
- The magic `$` variables still aren't sitting 100% right with me. I guess `$props` isn't too different from Astro.props, but something about adding a new `$data` one is now starting to feel like how 11ty/hugo/others do these magic global variables which I always find impossible to keep track of.
- use=case: Defining a default prop (a pretty common use-case) is now 1-step more indirect than what we have currently. Instead of setting a variable in the frontmatter and then using it directly in the template, we now have this layer of indirection where it needs to be translated through the `data()` fn. You then need to switch your template reference from `$props.foo` to `$data.foo` even though it's still really just a prop.

**Comment by FredKSchott** [comment]
One other, related idea that I was playing around with:

```jsx
---
export interface Props {
    title: string;
    description?: string[];
}
// Example 1: Setting a default prop
export function props(props: Props) => {
    return {
        description: 'This is the default site descrition.',
        ...props,
    };
}
// Example 2: Loading some data
export async function data(props: Props) {
    return {
        pokemon: await fetch('...'),
    };
}
---
<meta charset="UTF-8">
<title>{$props.title}</title>
<meta name="description" content={$props.description}>
<meta name="number-of-pokemon" content={$data.pokemon.length}>
```

#### What's different in this version:
- Introduces a `props()` fn specifically to handle the "setting a default prop" use-case.
- We could infer the result of `props()` with the `description` default applied to deliver on the idea of a second "internal" props type that includes the default prop value. This was one of the limits of the design above which forced us to move default props into the `data()` fn.
- I like the symmetry of `props()` and `data()` as the two functions, and then `$props` and `$data` as the two magic template variables. Even if its a bit more boilerplate, at least it's a really clear connection between the two. 

I think I like this one a bit more than the previous comment.

**Comment by matthewp** [comment]
I think the problems this proposal is trying to solve are small enough that they are not worth solving if there's a significant DX penalty for doing so. I think that would be the case if moving away from flat JS definitions; Vue is [moving towards](https://github.com/vuejs/rfcs/blob/master/active-rfcs/0040-script-setup.md) a flat definition as well.

Are these two things all that important?

1. Exporting from a .astro file.
2. Module scope data fetches.

I don't think they are, and if you need either case you can solve it yourself by defining a separate `.js` file to do, for example, data fetching, which will run only once.

---

I also disagree with this part of the motivation section:

> We would like to keep our scripting to a single block, since having multiple `---` blocks would break our current understanding of how frontmatter looks/feels.

I don't think this breaks understanding at all, because what we're really talking about here is a *corner case*. Most Svelte components do not have a setup script. The same way, if we had multiple `---` blocks in .astro files you would not need to use them in my cases. Only in the case where you do need some module-level scope; ie if you are data-fetching an API that might change frequently and you know the component is going to be used multiple times.

**Comment by natemoo-re** [comment]
I've been mulling this issue over. I think I agree with @matthewp, enabling exports from `.astro` files isn't a big enough issue to warrant a new mental model, so I'm generally -1 on any of these proposed solutions. I'd hate to cause more churn by changing the `props` syntax yet again.

The other issue (scoping of `async` requests) seems more important. Top-level await may be a bit of a footgun, but it's really straightforward and a great DX, so I don't think we should abandon it. My intuition is that shimming `node-fetch` to handle cache headers properly would solve the vast majority of use cases. 

Exposing lifecycle hooks (either `import { onSetup } from 'astro'` or `export async function onSetup()`) could also be a good solution. That would allow users to conditionally run some code based on how the component is being rendered. This seems very tied to the Collections API because you may want to load data for the entire collection or a single route of a collection.

**Comment by FredKSchott** [comment]
> I also disagree with this part of the motivation section:
> 
> > We would like to keep our scripting to a single block, since having multiple `---` blocks would break our current understanding of how frontmatter looks/feels.
> 
> I don't think this breaks understanding at all, because what we're really talking about here is a _corner case_. Most Svelte components do not have a setup script. The same way, if we had multiple `---` blocks in .astro files you would not need to use them in most cases. Only in the case where you do need some module-level scope; ie if you are data-fetching an API that might change frequently and you know the component is going to be used multiple times.

@matthewp really appreciate you calling this out! I agree that a second (or third, or forth) code fence isn't 100% out of the question forever, but I do think that it would be seen by the public as a break from "oh, this is just frontmatter" to "this is some special Astro syntax" and I think we're getting a lot of power from that argument today. I'd feel much more comfortable considering this out of scope for now and revisiting that idea in a v2 or v3 once we've gotten past our "new kid on the block" phase.

@matthewp @natemoo-re I'm not opposed to rejecting both of these proposals! Clarity is the big thing I'm looking for here. If I could summarize this third "do nothing" proposal as a flavor of OP:

1. The frontmatter script is always instance-level. (commitment to current behavior)
2. exports from an Astro component are not supported (you must export from some JS file instead)
3. To share state across components, use a store (you can also export a store)
4. Components are sync to import, async to render (commitment to current behavior)
    1. we support top-level await in the component syntax, but reminder that this run on every render
5. Perf considerations: we commit to protecting users from foot-gun perf issues with
    1. automatic warnings/debug help to identify slow components
    2. automatic respect fetch cache headers during dev/build
    3. memoization (`defineOnce`) and/or lifecycle (`onSetup`) helpers

Am I to understand that this currently has the most support? Did I get anything wrong or miss anything?

**Comment by FredKSchott** [comment]
I'm officially withdrawing this proposal for the moment. But I'll add one more important point that wasn't captured in the post above: if we keep supporting `export function createCollection` then we are already supporting a magic hoisting compiler. "but that's not how JavaScript works" was one of our main arguments against a magic compiler, but I think the fact that we have already crossed that line is important. At the same time, this is probably our most troublesome API, so there is good reason not to double down on it.

Also, to reiterate something that I mentioned in the RFC call: we could end up optimizing and bug-fixing the compiler that we have so much that we find ourselves with a magic hoisting compiler. In that case, I may resubmit this proposal in the future :)
